cluster-autoscaler/cloudprovider/aws/instance_type_cache_test.go

```diff
@@ -0,0 +1,182 @@
+			nil,
+		},
+		{
+			"AsgWithLaunchTemplateMixedInstancePolicyOverride",
+			nil,
+			nil,
+			&autoscaling.MixedInstancesPolicy{
+				LaunchTemplate: &autoscaling.LaunchTemplate{
+					Overrides: []*autoscaling.LaunchTemplateOverrides{
+						{
+							InstanceType: aws.String(instanceType),
+						},
+					},
+				},
+			},
+		},
+		{
+			"AsgWithLaunchTemplateMixedInstancePolicyNoOverride",
+			nil,
+			nil,
+			&autoscaling.MixedInstancesPolicy{
+				LaunchTemplate: &autoscaling.LaunchTemplate{
+					LaunchTemplateSpecification: &ltSpec,
+					Overrides:                   []*autoscaling.LaunchTemplateOverrides{},
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		m := newAsgInstanceTypeCache(&autoScalingWrapper{a}, &ec2Wrapper{e})
+
+		err := m.populate([]*autoscaling.Group{
+			{
+				AutoScalingGroupName:    aws.String(asgName),
+				LaunchConfigurationName: tc.launchConfigurationName,
+				LaunchTemplate:          tc.launchTemplate,
+				MixedInstancesPolicy:    tc.mixedInstancesPolicy,
+			},
+		})
+		assert.NoError(t, err)
+
+		result, found, err := m.GetByKey(asgName)
+		assert.NoErrorf(t, err, "%s had error %v", tc.name, err)
+		assert.Truef(t, found, "%s did not find asg", tc.name)
+
+		foundInstanceType := result.(instanceTypeCachedObject).instanceType
+		assert.Equalf(t, foundInstanceType, instanceType, "%s had %s, expected %s", tc.name, foundInstanceType, instanceType)
+	}
+}
+
+func TestLTVersionChange(t *testing.T) {
+	asgName, ltName := "testasg", "launcher"
+	ltVersions := []*string{aws.String("1"), aws.String("2")}
+	instanceTypes := []*string{aws.String("t2.large"), aws.String("m4.xlarge")}
+
+	a := &AutoScalingMock{}
+	e := &EC2Mock{}
+
+	for i := 0; i < 2; i++ {
+		e.On("DescribeLaunchTemplateVersions", &ec2.DescribeLaunchTemplateVersionsInput{
+			LaunchTemplateName: aws.String(ltName),
+			Versions:           []*string{ltVersions[i]},
+		}).Return(&ec2.DescribeLaunchTemplateVersionsOutput{
+			LaunchTemplateVersions: []*ec2.LaunchTemplateVersion{
+				{
+					LaunchTemplateData: &ec2.ResponseLaunchTemplateData{
+						InstanceType: instanceTypes[i],
+					},
+				},
+			},
+		})
+	}
+
+	ttl := time.Second * 2
+	m := newAsgInstanceTypeCacheWithTTL(&autoScalingWrapper{a}, &ec2Wrapper{e}, ttl)
+
+	for i := 0; i < 2; i++ {
+		err := m.populate([]*autoscaling.Group{
+			{
+				AutoScalingGroupName: aws.String(asgName),
+				LaunchTemplate: &autoscaling.LaunchTemplateSpecification{
+					LaunchTemplateName: aws.String(ltName),
+					Version:            ltVersions[i],
+				},
+			},
+		})
+		assert.NoError(t, err)
+
+		result, found, err := m.GetByKey(asgName)
+		assert.NoError(t, err)
+		assert.Truef(t, found, "%s did not find asg", asgName)
+
+		foundInstanceType := result.(instanceTypeCachedObject).instanceType
+		assert.Equalf(t, foundInstanceType, *instanceTypes[i], "%s had %s, expected %s", asgName, foundInstanceType, *instanceTypes[i])
+
+		// ensure the cache has expired
+		if i != 1 {
+			time.Sleep(ttl)
```

Comment:
Dont have to fix it here. (Since this is just moving an existing function call).

This introduces build failures and slows build by 4 seconds.
When the build runs on a slow machine, cache.Store thread might not get cycles to clear the cache and this test might fail?

Reply:
Fixed